Examples on sodapy can further be located here: https://github.com/xmunoz/sodapy/blob/master/examples/soql_queries.ipynb


In [8]:
import json

with open('credentials.json') as f:
    data = json.load(f)
    socrates_id = data['socrates_id']
    socrates_key = data['socrates_api_key']

In [3]:
import pandas as pd
from sodapy import Socrata
from IPython.core.display import display, HTML

display(HTML("<style>.container { width:100% !important; }</style>"))
pd.set_option('display.max_colwidth', None)

In [37]:
inventory_db = "6vkj-f5xf"
loans_db = "5src-czff"

client = Socrata("data.seattle.gov", None, username=socrates_id,password=socrates_key, timeout=50)

# Example authenticated client (needed for non-public datasets):
# client = Socrata(data.seattle.gov,
#                  MyAppToken,
#                  userame="user@example.com",
#                  password="AFakePassword")

# First 2000 results, returned as JSON from API / converted to Python list of
# dictionaries by sodapy.
results = client.get(inventory_db, limit=5000)

# Convert to pandas DataFrame
results_df = pd.DataFrame.from_records(results)
results_df.head()

bibnum  \
0  3011076   
1  2248846   
2  3209270   
3  1907265   
4  1644616   

                                                                                                                              title  \
0  A tale of two friends / adapted by Ellie O'Ryan ; illustrated by Tom Caulfield, Frederick Gardner, Megan Petasky, and Allen Tam.   
1                           Naruto. Vol. 1, Uzumaki Naruto / story and art by Masashi Kishimoto ; [English adaptation by Jo Duffy].   
2                                                           Peace, love & Wi-Fi : a ZITS treasury / by Jerry Scott and Jim Borgman.   
3                                                                                    The Paris pilgrims : a novel / Clancy Carlile.   
4                       Erotic by nature : a celebration of life, of love, and of our wonderful bodies / edited by David Steinberg.   

                      author  \
0              O'Ryan, Ellie   
1  Kishimoto, Masashi, 1974-   
2        Scott, Jerry, 1955-   
3     Carlile, Clancy, 1930-   
4                        NaN   

                                                   isbn publicationyear  \
0  1481425730, 1481425749, 9781481425735, 9781481425742           2014.   
1                                            1569319006    2003, c1999.   
2                             144945867X, 9781449458676           2014.   
3                                            0786706155          c1999.   
4                                            094020813X    1991, c1988.   

                           publisher  \
0                   Simon Spotlight,   
1                               Viz,   
2         Andrews McMeel Publishing,   
3                    Carroll & Graf,   
4  Red Alder Books/Down There Press,   

                                                                                                                                                                                                                                                subjects  \
0                                                                                                                                   Musicians Fiction, Bullfighters Fiction, Best friends Fiction, Friendship Fiction, Adventure and adventurers Fiction   
1                                                                                                                                             Ninja Japan Comic books strips etc, Comic books strips etc Japan Translations into English, Graphic novels   
2  Duncan Jeremy Fictitious character Comic books strips etc, Teenagers United States Comic books strips etc, Parent and teenager Comic books strips etc, Families Comic books strips etc, Comic books strips etc, Comics Graphic works, Humorous comics   
3                                                                                                                                                                           Hemingway Ernest 1899 1961 Fiction, Biographical fiction, Historical fiction   
4                                                                                                                                                                                           Erotic literature American, American literature 20th century   

  itemtype itemcollection floatingitem itemlocation               reportdate  \
0     jcbk          ncrdr     Floating          qna  2017-09-01T00:00:00.000   
1     acbk        nycomic           NA          lcy  2017-09-01T00:00:00.000   
2     acbk        nycomic           NA          bea  2017-09-01T00:00:00.000   
3     acbk          cafic           NA          cen  2017-09-01T00:00:00.000   
4     acbk           canf           NA          cen  2017-09-01T00:00:00.000   

  itemcount  
0         1  
1         1  
2         1  
3         1  
4         1

In [14]:
results = client.get(loans_db, limit=5000)

# Convert to pandas DataFrame
results_df = pd.DataFrame.from_records(results)
results_df.head()

id checkoutyear bibnumber    itembarcode itemtype  \
0  201403091424000010077167236         2014   2830415  0010077167236     acbk   
1  201501271027000010073337015         2015   2747196  0010073337015     jcbk   
2  201503211600000010079375852         2015   2289234  0010079375852     accd   
3  201509041150000010085400504         2015   3110065  0010085400504     accd   
4  201507171420000010085618998         2015   2928031  0010085618998     jcbk   

  collection                callnumber  \
0       nanf      616.8522 Sh224G 2012   
1      ncrdr                 ER CATROW   
2       nacd       CD 789.30944 P2183M   
3       nacd          CD 782.42166 X1U   
4    nccomic  J741.5952 K9681Pb10 2013   

                                                                                      itemtitle  \
0  Getting past your past take control of your life with self help techniques from EMDR therapy   
1                                                                                  Best in Show   
2                                                                                        Motifs   
3                                                                       Under the big black sun   
4                                                                Pokemon Black and White Vol 10   

                                                                                subjects  \
0             Eye movement desensitization and reprocessing, Anxiety disorders Treatment   
1                    Dog shows Juvenile fiction, Spaniels Juvenile fiction, Dogs Fiction   
2                                           Popular music France 2001 2010, Songs French   
3                                                                   Rock music 1981 1990   
4  Adventure and adventurers Comic books strips etc, Cartoons and comics, Graphic novels   

          checkoutdatetime  
0  2014-03-09T14:24:00.000  
1  2015-01-27T10:27:00.000  
2  2015-03-21T16:00:00.000  
3  2015-09-04T11:50:00.000  
4  2015-07-17T14:20:00.000

In [27]:
results = client.get(inventory_db, where='itemlocation="hip"', limit=100000)

# Convert to pandas DataFrame
results_df = pd.DataFrame.from_records(results)
results_df.head()

bibnum  \
0  3119028   
1  3481909   
2  3378040   
3  3355446   
4  3354738   

                                                                                                                        title  \
0                                   Notorious RBG : the life and times of Ruth Bader Ginsburg / Irin Carmon & Shana Knizhnik.   
1                Through the water curtain : & other tales from around the world / selected and introduced by Cornelia Funke.   
2  Miraculous : tales of Ladybug & Cat Noir. Bug out / adapted by Cheryl Black & Nicole D'Andria ; lettered by Justing Birch.   
3                                                                                                  Broken ice / Matt Goldman.   
4                                                                                Rainbow dash rights the ship / G. M. Berrow.   

                 author  \
0          Carmon, Irin   
1                   NaN   
2         Black, Cheryl   
3  Goldman, Matt, 1962-   
4         Berrow, G. M.   

                                                                              isbn  \
0  0062415824, 0062415832, 0062425714, 9780062415820, 9780062415837, 9780062425713   
1                                         1782692002, 9781782692003, 9781782692010   
2                                                        1632293668, 9781632293664   
3                                                        0765391317, 9780765391315   
4                             0316557528, 1536433055, 9780316557528, 9781536433050   

  publicationyear                                                   publisher  \
0          [2015]  Dey Street Books, An Imprint of William Morrow Publishers,   
1           2018.                                         Pushkin Children's,   
2          [2018]                                   Action Lab Entertainment,   
3           2018.                                                      Forge,   
4           2018.                                  Little, Brown and Company,   

                                                                                                                                                subjects  \
0                                                                                                    Ginsburg Ruth Bader, Judges United States Biography   
1                                                                                                                           Tales, Folklore, Fairy tales   
2  Teenage girls Comic books strips etc, Superheroes Comic books strips etc, Cartoons and comics, Graphic novels, Comics Graphic works, Superhero comics   
3                         Private investigators Fiction, Missing persons Fiction, Minnesota Fiction, Warroad Minn Fiction, Detective and mystery fiction   
4                                                                 Ponies Juvenile fiction, Friendship Juvenile fiction, Fantasy fiction, Fantasy fiction   

  itemtype itemcollection floatingitem itemlocation               reportdate  \
0     acbk            nab     Floating          hip  2020-05-01T00:00:00.000   
1     jcbk          ncfft           NA          hip  2020-05-01T00:00:00.000   
2     jcbk        nccomic     Floating          hip  2020-05-01T00:00:00.000   
3     acbk          namys     Floating          hip  2020-05-01T00:00:00.000   
4     jcbk          ncfic           NA          hip  2020-05-01T00:00:00.000   

  itemcount  
0         1  
1         1  
2         1  
3         1  
4         1

In [28]:
len(results_df)

100000

In [29]:
results = client.get(inventory_db, where='itemlocation="hip"', limit=100)

# Convert to pandas DataFrame
results_df = pd.DataFrame.from_records(results)
results_df.head()

bibnum  \
0  3119028   
1  3481909   
2  3378040   
3  3355446   
4  3354738   

                                                                                                                        title  \
0                                   Notorious RBG : the life and times of Ruth Bader Ginsburg / Irin Carmon & Shana Knizhnik.   
1                Through the water curtain : & other tales from around the world / selected and introduced by Cornelia Funke.   
2  Miraculous : tales of Ladybug & Cat Noir. Bug out / adapted by Cheryl Black & Nicole D'Andria ; lettered by Justing Birch.   
3                                                                                                  Broken ice / Matt Goldman.   
4                                                                                Rainbow dash rights the ship / G. M. Berrow.   

                 author  \
0          Carmon, Irin   
1                   NaN   
2         Black, Cheryl   
3  Goldman, Matt, 1962-   
4         Berrow, G. M.   

                                                                              isbn  \
0  0062415824, 0062415832, 0062425714, 9780062415820, 9780062415837, 9780062425713   
1                                         1782692002, 9781782692003, 9781782692010   
2                                                        1632293668, 9781632293664   
3                                                        0765391317, 9780765391315   
4                             0316557528, 1536433055, 9780316557528, 9781536433050   

  publicationyear                                                   publisher  \
0          [2015]  Dey Street Books, An Imprint of William Morrow Publishers,   
1           2018.                                         Pushkin Children's,   
2          [2018]                                   Action Lab Entertainment,   
3           2018.                                                      Forge,   
4           2018.                                  Little, Brown and Company,   

                                                                                                                                                subjects  \
0                                                                                                    Ginsburg Ruth Bader, Judges United States Biography   
1                                                                                                                           Tales, Folklore, Fairy tales   
2  Teenage girls Comic books strips etc, Superheroes Comic books strips etc, Cartoons and comics, Graphic novels, Comics Graphic works, Superhero comics   
3                         Private investigators Fiction, Missing persons Fiction, Minnesota Fiction, Warroad Minn Fiction, Detective and mystery fiction   
4                                                                 Ponies Juvenile fiction, Friendship Juvenile fiction, Fantasy fiction, Fantasy fiction   

  itemtype itemcollection floatingitem itemlocation               reportdate  \
0     acbk            nab     Floating          hip  2020-05-01T00:00:00.000   
1     jcbk          ncfft           NA          hip  2020-05-01T00:00:00.000   
2     jcbk        nccomic     Floating          hip  2020-05-01T00:00:00.000   
3     acbk          namys     Floating          hip  2020-05-01T00:00:00.000   
4     jcbk          ncfic           NA          hip  2020-05-01T00:00:00.000   

  itemcount  
0         1  
1         1  
2         1  
3         1  
4         1

In [35]:
results = client.get(inventory_db, where='itemlocation="hip"', limit=100, offset=1)

# Convert to pandas DataFrame
results_df = pd.DataFrame.from_records(results)
results_df.head()

bibnum  \
0  3481909   
1  3378040   
2  3355446   
3  3354738   
4   634098   

                                                                                                                        title  \
0                Through the water curtain : & other tales from around the world / selected and introduced by Cornelia Funke.   
1  Miraculous : tales of Ladybug & Cat Noir. Bug out / adapted by Cheryl Black & Nicole D'Andria ; lettered by Justing Birch.   
2                                                                                                  Broken ice / Matt Goldman.   
3                                                                                Rainbow dash rights the ship / G. M. Berrow.   
4                                                                The listening walk / by Paul Showers ; illustrated by Aliki.   

                                                   isbn publicationyear  \
0              1782692002, 9781782692003, 9781782692010           2018.   
1                             1632293668, 9781632293664          [2018]   
2                             0765391317, 9780765391315           2018.   
3  0316557528, 1536433055, 9780316557528, 9781536433050           2018.   
4                 0060216379, 0060216387, 9780064433228           1991.   

                    publisher  \
0         Pushkin Children's,   
1   Action Lab Entertainment,   
2                      Forge,   
3  Little, Brown and Company,   
4              HarperCollins,   

                                                                                                                                                subjects  \
0                                                                                                                           Tales, Folklore, Fairy tales   
1  Teenage girls Comic books strips etc, Superheroes Comic books strips etc, Cartoons and comics, Graphic novels, Comics Graphic works, Superhero comics   
2                         Private investigators Fiction, Missing persons Fiction, Minnesota Fiction, Warroad Minn Fiction, Detective and mystery fiction   
3                                                                 Ponies Juvenile fiction, Friendship Juvenile fiction, Fantasy fiction, Fantasy fiction   
4                                                                                                  Fathers and daughters Juvenile fiction, Sound Fiction   

  itemtype itemcollection floatingitem itemlocation               reportdate  \
0     jcbk          ncfft           NA          hip  2020-05-01T00:00:00.000   
1     jcbk        nccomic     Floating          hip  2020-05-01T00:00:00.000   
2     acbk          namys     Floating          hip  2020-05-01T00:00:00.000   
3     jcbk          ncfic           NA          hip  2020-05-01T00:00:00.000   
4     jcbk          ncpic           NA          hip  2020-05-01T00:00:00.000   

  itemcount                author  
0         1                   NaN  
1         1         Black, Cheryl  
2         1  Goldman, Matt, 1962-  
3         1         Berrow, G. M.  
4         1         Showers, Paul

In [32]:
meta = client.get_metadata(inventory_db)
meta

{'id': '6vkj-f5xf',
 'name': 'Library Collection Inventory',
 'attribution': 'City of Seattle',
 'attributionLink': 'http://www.spl.org/',
 'averageRating': 0,
 'category': 'Community',
 'createdAt': 1503607716,
 'description': 'The Seattle Public Library\'s collection inventory. \n\nThere is a data dictionary available to help you understand the codes used in the "ItemCollection", "ItemType" and "ItemLocation" columns. Access here:\nhttps://data.seattle.gov/Community/Integrated-Library-System-ILS-Data-Dictionary/pbt3-ytbc',
 'displayType': 'table',
 'downloadCount': 582,
 'hideFromCatalog': False,
 'hideFromDataJson': False,
 'indexUpdatedAt': 1546789412,
 'licenseId': 'PUBLIC_DOMAIN',
 'newBackend': True,
 'numberOfComments': 0,
 'oid': 26657588,
 'provenance': 'official',
 'publicationAppendEnabled': False,
 'publicationDate': 1507302241,
 'publicationGroup': 14420133,
 'publicationStage': 'published',
 'rowClass': 'http://www.w3.org/ns/dcat#Dataset',
 'rowsUpdatedAt': 1588767753,
 

In [39]:
results = client.get(inventory_db, select="itemlocation, count(*)", group = "itemlocation")

# Convert to pandas DataFrame
results_df = pd.DataFrame.from_records(results)
results_df.head()

itemlocation    count
0          bal  1713293
1          BAL     1747
2          bea  1150576
3          BEA      128
4          bro  1374348

In [40]:
results_df

itemlocation     count
0           bal   1713293
1           BAL      1747
2           bea   1150576
3           BEA       128
4           bro   1374348
5           BRO        78
6           cap   1168103
7           CAP        44
8           cen  18992386
9           col   1035716
10          COL         3
11          dlr    487608
12         drp1       165
13          dth   1515314
14          fre    541601
15          glk    832663
16          gwd   1664790
17          GWD      7122
18          hip    643319
19          idc    375510
20          ill        21
21          lcy   1466050
22          mag    820389
23          MAG        38
24          mgm    337203
25          mob    152859
26          mon    535845
27          MON        39
28          net   1961060
29          NET      2625
30          nga   1116911
31          nhy    352942
32          out        38
33          qna    762414
34          QNA        19
35          rbe    991785
36          RBE       332
37          spa    503619
38         stak         1
39          swt   1579122
40          tcs     20782
41          uni    946931
42          wal    276526
43          wlb         3
44          wts   1053725
45          WTS       537

can consider mob (Mobile Library)
mob 	152859

based on seattle library website: https://www.spl.org/programs-and-services/outreach/mobile-library-services

This drives home the need for better book matching